### PR TITLE
Class21/week2

### PR DESCRIPTION
**Pull request by ghifo**

### CODE REVIEW

File: week2/homework/bonus/bonus.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
+const program = require('commander');
+const fs = require('fs');
+
+program
+  .option('add <new todo>', 'Add new todo item', function(todoItem) {
+    fs.readFile('./bonus/list.json', 'utf8', (error, toDos) => {
+      if (error) {
+        console.log(error);
+      }
+      const todoList = JSON.parse(toDos);
```
Comment (gajduk) [inline week2/homework/bonus/bonus.js]: use of json 👍

File: week2/homework/bonus/bonus.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
+const program = require('commander');
```
Comment (gajduk) [inline week2/homework/bonus/bonus.js]: commander 👍

File: week2/homework/bonus/bonus.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        console.log(error);
+      }
+      const todoList = JSON.parse(toDos);
+      const filteredList = todoList.filter((elem, index) => index !== Number(todoItem) - 2);
+      const newList = JSON.stringify(filteredList, null, 2);
+      console.log(newList);
+      return fs.writeFile('./bonus/list.json', newList, error => console.log(error));
+    });
+  })
+  // The `update` takes only one argument, the second one is `undefined`.
```
Comment (gajduk) [inline week2/homework/bonus/bonus.js]: Looks like a known issue. It is suggested to use `program.args` https://github.com/tj/commander.js/issues/53#issuecomment-59824905

File: week2/homework/bonus/bonus.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  // When I try to update an item it becomes `null`
+  .option('update <row number> <new value>', 'update one todo item', function(todoItem, newVal) {
+    fs.readFile('./bonus/list.json', 'utf8', (error, toDos) => {
+      if (error) {
+        console.log(error);
+      }
+      const todoList = JSON.parse(toDos);
+      const mappedList = todoList.map((elem, index) =>
+        index === Number(todoItem) - 2 ? newVal : elem,
```
Comment (gajduk) [inline week2/homework/bonus/bonus.js]: Why - 2 ?

File: week2/homework/src/index.js [hunk 5]
```diff
@@ -1,3 +1,34 @@
 'use strict';
+const args = process.argv.slice(2);
+const command = args[0];
+const todoItem = args[1];
+const newVal = args[2];
 
-// TODO: Write the homework code in this file
+const help = require('./help');
```
Comment (gajduk) [inline week2/homework/src/index.js]: separate files for each command 👍

File: week2/homework/src/index.js [hunk 6]
```diff
@@ -1,3 +1,34 @@
...
+const remove = require('./remove');
+const reset = require('./reset');
+const update = require('./update');
+
+switch (command) {
+  case 'list':
+    list().then(toDos => console.log(toDos));
+    break;
+  case 'add':
+    add(todoItem).then(toDos => console.log(toDos));
```
Comment (gajduk) [inline week2/homework/src/index.js]: nice use of promises 👍

File: week2/homework/src/list.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+const fs = require('fs');
+const fileName = './src/data.json';
+
+function list() {
+  return new Promise((resolve, reject) => {
+    fs.readFile(fileName, 'utf8', (error, toDos) => {
+      error ? reject(error) : resolve(toDos);
```
Comment (gajduk) [inline week2/homework/src/list.js]: error handling 👍

File: week2/homework/src/update.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
+const fs = require('fs');
+const list = require('./list');
+const help = require('./help');
+const fileName = './src/data.json';
+
+function update(todoItem, newVal) {
+  return new Promise((resolve, reject) => {
+    list().then(toDos => {
```
Comment (gajduk) [inline week2/homework/src/update.js]: what if list throws an error?

File: week2/homework/src/list.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
+const fs = require('fs');
+const fileName = './src/data.json';
+
+function list() {
```
Comment (gajduk) [inline week2/homework/src/list.js]: separate function for loading the file 👍

File: week2/homework/bonus/bonus.js (hunk 2 above)
Comment (ghifo) [reply week2/homework/bonus/bonus.js]: It was very new, I hope I could do it properly.

File: week2/homework/bonus/bonus.js (hunk 4 above)
Comment (ghifo) [reply week2/homework/bonus/bonus.js]: To reach `todoItem` by its line number. The first line of the `json` file is `[` and the first `todoItem` is line `2`

File: week2/homework/src/index.js (hunk 6 above)
Comment (ghifo) [reply week2/homework/src/index.js]: Thanks!

File: week2/homework/src/update.js (hunk 8 above)
Comment (ghifo) [reply week2/homework/src/update.js]: The `list()` handles `errors` itself and the `writeFile` handles it as well, is not that enough?!

File: week2/homework/bonus/bonus.js (hunk 3 above)
Comment (ghifo) [reply week2/homework/bonus/bonus.js]: I tried this one but it didn't work at all for some reason.

program
  .command('update')
  .usage('<row number> [new value...]')
  .description('update one todo item')
  .action(function(todoItem, newVal) {
    fs.readFile('./bonus/list.json', 'utf8', (error, toDos) => {
      if (error) {
        console.log(error);
      }
      const todoList = JSON.parse(toDos);
      const mappedList = todoList.map((elem, index) =>
        index === Number(todoItem) - 2 ? newVal : elem,
      );
      const newList = JSON.stringify(mappedList, null, 2);
      console.log(newList);
      return fs.writeFile('./bonus/list.json', newList, error => console.log(error));
    });
  });
`node bonus . update 2 "Do something else"` 
It does not make any changes and no messages at all in the `console`

Comment (gajduk) [reply week2/homework/bonus/bonus.js]: hmm, well no big deal, it is not your fault, it is an error in the library

File: week2/homework/bonus/bonus.js (hunk 4 above)
Comment (gajduk) [reply week2/homework/bonus/bonus.js]: But you are using the index for the json array not for the line number in the file

File: week2/homework/src/update.js (hunk 8 above)
Comment (gajduk) [reply week2/homework/src/update.js]: list() rethrows the error, so it needs to be handled here

File: week2/homework/bonus/bonus.js (hunk 4 above)
Comment (ghifo) [reply week2/homework/bonus/bonus.js]: So `-1` will be enough this way.
I'll do that.

File: week2/homework/src/update.js (hunk 8 above)
Comment (ghifo) [reply week2/homework/src/update.js]: So I have to use the `.catch` to solve it

Comment (ghifo) [reply week2/homework/src/update.js]: `.catch(error => console.log(error));`

File: week2/homework/bonus/bonus.js (hunk 3 above)
Comment (ghifo) [reply week2/homework/bonus/bonus.js]: Great!

**Pull request closed without merging**